src/libcollections/btree/map.rs

```diff
@@ -463,6 +464,142 @@ impl<K: Ord, V> BTreeMap<K, V> {
             }
         }
     }
+
+    /// Moves all elements from `other` into `Self`, leaving `other` empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #![feature(btree_append_split_off)]
+    /// use std::collections::BTreeMap;
+    ///
+    /// let mut a = BTreeMap::new();
+    /// a.insert(1, "a");
+    /// a.insert(2, "b");
+    /// a.insert(3, "c");
+    ///
+    /// let mut b = BTreeMap::new();
+    /// b.insert(3, "d");
+    /// b.insert(4, "e");
+    /// b.insert(5, "f");
+    ///
+    /// a.append(&mut b);
+    ///
+    /// assert_eq!(a.len(), 5);
+    /// assert_eq!(b.len(), 0);
+    ///
+    /// assert_eq!(a[&1], "a");
+    /// assert_eq!(a[&2], "b");
+    /// assert_eq!(a[&3], "d");
+    /// assert_eq!(a[&4], "e");
+    /// assert_eq!(a[&5], "f");
+    /// ```
+    #[unstable(feature = "btree_append_split_off",
+               reason = "recently added as part of collections reform 2")]
+    pub fn append(&mut self, other: &mut Self) {
+        // Read all values from `other` into `self`.
+        // The use of `ptr::read` is safe as we clear `other` afterwards
+        for (key, value) in other.iter() {
+            unsafe {
+                self.insert(ptr::read(key), ptr::read(value));
+            }
+        }
+        other.clear();
+    }
+
+    /// Splits the map into two at the given key,
+    /// retaining the first half in-place and returning the second one.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #![feature(btree_append_split_off)]
+    /// use std::collections::BTreeMap;
+    ///
+    /// a.insert(1, "a");
+    /// a.insert(2, "b");
+    /// a.insert(3, "c");
+    /// a.insert(4, "d");
+    /// a.insert(5, "e");
+    ///
+    /// let b = a.split_off(3);
+    ///
+    /// assert_eq!(a.len(), 2);
+    /// assert_eq!(b.len(), 3);
+    ///
+    /// assert_eq!(a[&1], "a");
+    /// assert_eq!(a[&2], "b");
+    /// assert_eq!(b[&3], "c");
+    /// assert_eq!(b[&4], "d");
+    /// assert_eq!(b[&5], "e");
+    /// ```
+    #[unstable(feature = "btree_append_split_off",
+               reason = "recently added as part of collections reform 2")]
+    pub fn split_off<B: Borrow<K>>(&mut self, at: B) -> Self {
+        let mut other = BTreeMap::new();
+
+        if self.len() == 0 {
+            return other;
+        }
+
+        let at = at.borrow();
+
+        // FIXME(RFC #811) We can't check for `at` pointing before the
+        // first element and then swap `self` and `other`, because
+        // `self` will still be borrowed immutably.
+        let mut should_swap = false;
+        {
+            // Using `unwrap` is safe because `self.len()` > 0.
+            if at <= self.keys().next().unwrap() {
```

Comment:
Documenting unwraps like this is great, but I would avoid using “safe”. Panic is well defined and memory safe, and there's a lot of unsafe blocks in the file, so it's better to avoid the word safe when it's not about memory safety.